### PR TITLE
Extract IPC ArrayReader struct

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Relates to #4255

# Rationale for this change
 

Whilst investigating #4255 I realised this code has naturally grown to the point of it now being quite hard to follow. This is the first of what will probably be a series of PRs to 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-ipc/src/reader.rs [hunk 1]
```diff
@@ -334,125 +205,21 @@ fn create_array(
...
-            array
-        }
-    };
-    Ok((array, node_index, buffer_index))
-}
-
-/// Skip fields based on data types to advance `node_index` and `buffer_index`.
-/// This function should be called when doing projection in fn `read_record_batch`.
-/// The advancement logic references fn `create_array`.
-fn skip_field(
```
Comment (tustvold) [inline arrow-ipc/src/reader.rs:367]: This logic is moved into skip_node

File: arrow-ipc/src/reader.rs [hunk 2]
```diff
@@ -628,14 +395,108 @@ fn create_dictionary_array(
...
+    fn next_node(&mut self, field: &Field) -> Result<&'a FieldNode, ArrowError> {
+        self.nodes.next().ok_or_else(|| {
+            ArrowError::IoError(format!(
+                "Invalid data for schema. {} refers to node not found in schema",
+                field
+            ))
+        })
+    }
+
+    fn skip_node(&mut self, field: &Field) -> Result<(), ArrowError> {
```
Comment (tustvold) [inline arrow-ipc/src/reader.rs:432]: This moved from skip_field

**Comment by viirya** [comment]
I'll review this soon.

**Review by viirya — APPROVED** [review]
Compared previous and new state. Looks good to me. It is more understandable to others, I think. 👍

File: arrow-ipc/src/reader.rs (hunk 2 above)
Comment (viirya) [reply arrow-ipc/src/reader.rs:432]: I feel original name `skip_field` is more proper, but okay for `skip_node`.
